kizhin.evgeniy/P3

sdjklfjsdlkfjsdklfj opened the pull request:
CNT-LOC-MAX

Commit:
Initial commit for P3

Created 3 modules: matrix_utils, io_utils, parse_arguments
Created main

Commit:
Create enums ExitCodes and MemoryModes

Added enums to avoid using "magic numbers"

Commit:
Refactor: restructure project and memory handling

- Removed exit_codes.hpp: replaced usage with in-scope enum
  in main.cpp.
- Refactored parse_arguments module:
  - Used checkArgumentsCount function and parseArguments which
    will call directly from main withot mediator functions such as
    parseArguments that was removed
  - Eliminated unnecessary error message streams, replaced with
    exception handling in main.
- Refactored memory_utils:
  - Renamed memory_modes.hpp to memory_utils.hpp.
  - Centralized allocateArray and deallocateArray functions in
    memory_utils.cpp, allowing memory management and MemoryMode
    definitions in a single module.
- Refined main.cpp logic:
  - Moved file handling to structured error handling, replacing
    inline dimension reading in io_utils.
  - Improved readability by inlining error handling for file open
    and data read/write operations.
- Removed io_utils:
  - Short functions such as readMatrixDimensions, writeResultToFile
    were inlined into main
  - Function readArrayValues were moved to matrix_utils module

- Added null and zero checks in functions allocateArray,
  countLocalMinimums, readArrayValues
- Other minor changes

Commit:
Update matrix_utils module

- Refactor argument checkings
- Renamed varibale neighborsConunt -> neighborsSize
- Renamed varibale tmp -> center

Commit:
Update parse_arguments module

- Fiexed typo (parseProgrammMode -> parseProgramMode)
- Added varibale for expected arguments size
- Fixed error massages

Commit:
Update memory_utils module

- Added size check before stack allocation
- Used exceptions instead of returning nullptr

Commit:
Update main.cpp

- Changed headers order accoring to CG
- Block with files input/output was moved to try-catch to avoid
  repeating return operator, deallocation and std::cerr using

Commit:
Update memory_utils and matrix_utils

- Removed null value from MemoryMode enum
- Removed unnecessary rows and columns variables: to avoid code
  repetition, matrix now stores rows and columns in the first two
  cells. This allows functions to pass only matrix pointers
  without size values.
- Created function initializeMatrix in matrix_utils to separate
  matrix creation and reading logic from main.

Commit:
Update memory handling

Used automatic storage instead of static storage according to
task T-T

Commit:
Refactor code style

- Renamed function from parse_arguments module: parseProgramMode->
  parseMemoryMode
- Renamed file paths variables in main: fileIn -> inputFilePath...
- Removed function initializeMatrix from matrix_utils (it had too
  much responibility). Use directily calls of other function and
  operations in main instead
- Removed ExitCodes enum from main (it might be better without but
  I'm not quite sure)
- Removed close file objects calls from main (it's uneccessary because
  of a probability of exceptions (but I'm not quite sure)
- Added check to readArrayValues function (returns if smth wrong
  with stream)
- Refactored variable names in parseMemoryMode function
- Updated function isLocalMinimum to check all 8 cells
- Minor code style changes

Commit:
Refactor main: stream managment, code style

- Renamed some variables
- Moved stackMemorySize to memory_utils header (it used be both
  main and memory_utils so better make it a global constant)
- Closed streams in main (idk is it right decision)
- Minor changes
- Used .good() stream method to check is reading passed correctly
  instead of just bool convertion because it allows to check eof
  state (idk why but previous variant worked corrently as well..)
- Used convertion stream objects to bool to check file openning
  (it should be the same but less uneccessary letters)
- Remove close methods because "This is what RAII is for, let the
  destructor do its job. There is no harm in closing it manually,
  but it's not the C++ way"

Commit:
Update main

Commit:
Update main

Volgarenok commented on kizhin.evgeniy/P3/matrix_utils.cpp:
```diff
@@ -0,0 +1,59 @@
+#include "matrix_utils.hpp"
+
+std::istream& kizhin::readArrayValues(std::istream& in, int* array, size_t size)
+{
+  for (size_t i = 0; i < size; ++i) {
+    if (!(in >> array[i])) {
+      return in;
+    }
+  }
+  return in;
+}
```
Ну давайте сделаем, чтоб количество прочитанных символов всё-таки возвращалось

Volgarenok commented on kizhin.evgeniy/P3/parse_arguments.cpp:
```diff
@@ -0,0 +1,27 @@
+  if (firstArgument == nullptr || firstArgument[0] == '\0' || firstArgument[1] != '\0') {
+    throw std::logic_error("Memory mode must be a single character");
+  }
+  if (firstArgument[0] == '1') {
+    return MemoryMode::stack;
+  }
+  if (firstArgument[0] == '2') {
+    return MemoryMode::freeStore;
+  }
```
А если там `"1 abc"`?

sdjklfjsdlkfjsdklfj replied:
Перед проверкой значения firstArgument есть дополнительная проверка, которая гарантирует, что firstArgument не является nullptr, не пустой строкой и содержит только один символ. Вот соответствующий код:

``` cpp
if (firstArgument == nullptr || firstArgument[0] == '\0' || firstArgument[1] != '\0') {
    throw std::logic_error("Memory mode must be a single character");
}
```

Eсли firstArgument содержит что-то вроде "1 abc", то условие firstArgument != '\0' сработает и будет выброшено исключение

Volgarenok commented on kizhin.evgeniy/P3/memory_utils.hpp:
```diff
@@ -0,0 +1,13 @@
+
+namespace kizhin {
+  constexpr size_t stackBufferSize = 10'000;
+  enum class MemoryMode { stack = 1, freeStore = 2 };
```
Ну не надо в одну строчку

Volgarenok commented on kizhin.evgeniy/P3/memory_utils.cpp:
```diff
@@ -0,0 +1,24 @@
+#include "memory_utils.hpp"
+#include <stdexcept>
+
+int* kizhin::allocateArray(size_t size, int* stackBuffer, MemoryMode mode)
+{
+  if (size == 0) {
+    return nullptr;
+  }
+  if (mode == MemoryMode::freeStore) {
+    return new int[size];
+  }
+  if (mode == MemoryMode::stack && size <= stackBufferSize) {
+    return stackBuffer;
+  }
+  throw std::logic_error("Invalid memory allocation request");
+}
+
+void kizhin::deallocateArray(int* matrix, MemoryMode mode)
+{
+  if (mode == MemoryMode::freeStore) {
+    delete[] matrix;
+  }
+}
```
Не очень удачное решение в целом. Так получается, что одни и те же условия проверяются всё время. У вас есть общая часть: получить как-то указатель на данные, вызвать для него нужные функции, потом что-то сделать в конце

Commit:
Update memory management

- Removed functions allocate/deallocate, used checks and directly calls of new/delete in main (idk why)
- Updated readArray functions, now it returns amount of read elements (idk why)
- Added comma in enum in favor of readability

Volgarenok commented on kizhin.evgeniy/P3/main.cpp:
```diff
@@ -0,0 +1,55 @@
+    }
+    out << kizhin::countLocalMaximums(matrix);
+  } catch (const std::exception& ex) {
+    if (memoryMode == kizhin::MemoryMode::freeStore) {
+      delete[] matrix;
+    }
```
Если вы заведёте отдельный указатель для динамической памяти, то вам не нужна будет эта проверка. Вы сможете всегда вызывать `delete` для этого указателя

Commit:
Update main: add new poiter for heap

Added pointer memoryFreeStore that allows to deallocate memory
without extra checks

Commit:
Merge branch 'master' into kizhin.evgeniy/P3
